core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMoveSelectorFactory.java

```
@@ -0,0 +1,68 @@
+
+    private static final int DEFAULT_MINIMUM_K = 2;
+    private static final int DEFAULT_MAXIMUM_K = 4;
+
+    public KOptListMoveSelectorFactory(KOptListMoveSelectorConfig moveSelectorConfig) {
+        super(moveSelectorConfig);
+    }
+
+    @Override
+    protected MoveSelector<Solution_> buildBaseMoveSelector(HeuristicConfigPolicy<Solution_> configPolicy,
+            SelectionCacheType minimumCacheType, boolean randomSelection) {
+        SelectionOrder selectionOrder = SelectionOrder.fromRandomSelectionBoolean(randomSelection);
+        EntitySelectorConfig entitySelectorConfig = new EntitySelectorConfig();
+        // TODO let's not forget this here
+        //entitySelectorConfig.setCacheType(SelectionCacheType.PHASE);
+        EntitySelector<Solution_> entitySelector =
+                EntitySelectorFactory.<Solution_> create(entitySelectorConfig)
+                        .buildEntitySelector(configPolicy, minimumCacheType, selectionOrder);
+        // TODO support coexistence of list and basic variables https://issues.redhat.com/browse/PLANNER-2755
+        GenuineVariableDescriptor<Solution_> variableDescriptor =
+                getTheOnlyVariableDescriptor(entitySelector.getEntityDescriptor());
+        if (!variableDescriptor.isListVariable()) {
+            throw new IllegalArgumentException("The subListSwapMoveSelector (" + config
+                    + ") can only be used when the domain model has a list variable."
+                    + " Check your @" + PlanningEntity.class.getSimpleName()
+                    + " and make sure it has a @" + PlanningListVariable.class.getSimpleName() + ".");
+        }
+
+        ValueSelector<Solution_> valueSelector = buildEntityDependentValueSelector(configPolicy,
+                entitySelector.getEntityDescriptor(), minimumCacheType, selectionOrder);
+        int minimumK = Objects.requireNonNullElse(config.getMinimumK(), DEFAULT_MINIMUM_K);
```

Let's fail fast when the minimum < 2.